File: src/main/java/com/example/finsentinel/service/news/NewsSentimentService.java
package com.example.finsentinel.service.news;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.model.ChatModel;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.stereotype.Service;

@Service
@Slf4j
@RequiredArgsConstructor
public class NewsSentimentService {

    private final ChatModel chatModel;

    /**
     * Classifies news sentiment using the LLM.
     * Returns "POSITIVE", "NEGATIVE", or "NEUTRAL".
     * Defaults to "NEUTRAL" on any failure to never block the enrichment pipeline.
     */
    public String classify(String title, String summary) {
        try {
            String text = title + (summary != null ? " " + summary : "");
            String promptText = "Classify the sentiment of this financial news as exactly one word: POSITIVE, NEGATIVE, or NEUTRAL. " +
                    "Only respond with one of those three words, nothing else.\n\nNews: " + text;

            String response = chatModel.call(new Prompt(promptText))
                    .getResult()
                    .getOutput()
                    .getText()
                    .trim()
                    .toUpperCase();

            if (response.contains("POSITIVE")) return "POSITIVE";
            if (response.contains("NEGATIVE")) return "NEGATIVE";
            return "NEUTRAL";
        } catch (Exception e) {
            log.warn("Sentiment classification failed, defaulting to NEUTRAL: {}", e.getMessage());
            return "NEUTRAL";
        }
    }
}

File: apps/api/src/news/news-sentiment.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  DEFAULT_OPENROUTER_BASE_URL,
  DEFAULT_OPENROUTER_TEXT_MODEL,
} from '@finsentinel/ai-runtime';

/**
 * Sentiment classification result.
 */
export type Sentiment = 'POSITIVE' | 'NEGATIVE' | 'NEUTRAL';

const VALID_SENTIMENTS = new Set<string>(['POSITIVE', 'NEGATIVE', 'NEUTRAL']);

const SENTIMENT_PROMPT =
  'Classify the financial sentiment of this article as POSITIVE, NEGATIVE, or NEUTRAL. Respond with only one word.';

function normalizeOpenAIBaseUrl(value: string): string {
  const trimmed = value.replace(/\/$/, '');
  return trimmed.endsWith('/v1') ? trimmed : `${trimmed}/v1`;
}

/**
 * Uses an LLM via the configured OpenAI-compatible API to classify
 * the financial sentiment of a news article.
 *
 * Input: title + summary
 * Output: "POSITIVE" | "NEGATIVE" | "NEUTRAL"
 */
@Injectable()
export class NewsSentimentService {
  private readonly logger = new Logger(NewsSentimentService.name);
  private readonly apiKey: string;
  private readonly model: string;
  private readonly baseUrl: string;

  constructor(configService: ConfigService) {
    this.apiKey =
      configService.get<string>('ai.apiKey') || configService.get<string>('OPENROUTER_API_KEY', '');
    this.model =
      configService.get<string>('ai.model') ||
      configService.get<string>('AI_MODEL', DEFAULT_OPENROUTER_TEXT_MODEL);
    this.baseUrl = normalizeOpenAIBaseUrl(
      configService.get<string>('ai.baseUrl') ||
        configService.get<string>('OPENROUTER_BASE_URL', DEFAULT_OPENROUTER_BASE_URL),
    );
  }

  /**
   * Classify the financial sentiment of an article.
   *
   * @param title   - Article headline
   * @param summary - Article summary / description
   * @returns "POSITIVE" | "NEGATIVE" | "NEUTRAL"
   */
  async classify(title: string, summary: string | null): Promise<Sentiment> {
    const content = summary ? `Title: ${title}\nSummary: ${summary}` : `Title: ${title}`;

    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          messages: [
            { role: 'system', content: SENTIMENT_PROMPT },
            { role: 'user', content },
          ],
          max_tokens: 10,
          temperature: 0,
        }),
      });

      if (!response.ok) {
        const text = await response.text();
        this.logger.error(`Sentiment API error: ${response.status} — ${text}`);
        return 'NEUTRAL';
      }

      const data = (await response.json()) as {
        choices?: Array<{ message?: { content?: string } }>;
      };

      const rawAnswer = data.choices?.[0]?.message?.content?.trim().toUpperCase() ?? '';

      if (VALID_SENTIMENTS.has(rawAnswer)) {
        return rawAnswer as Sentiment;
      }

      this.logger.warn(`Unexpected sentiment response: "${rawAnswer}", defaulting to NEUTRAL`);
      return 'NEUTRAL';
    } catch (err) {
      this.logger.error(
        `Sentiment classification failed: ${err instanceof Error ? err.message : String(err)}`,
      );
      return 'NEUTRAL';
    }
  }
}
